### PR TITLE
Added the binding for TwitterImagePipeline v2.2.2

### PR DESCRIPTION
**Pull request by mattleibow**
- [x] still need to create component images

### CODE REVIEW

**Comment by mattleibow** [comment]
I am having issues binding this one:

```csharp
// @protocol TIPLogger <NSObject>
[Protocol, Model]
[BaseType(typeof(NSObject), Name = "TIPLogger")]
interface TIPLogger
{
	// @required -(void)tip_logWithLevel:(TIPLogLevel)level file:(NSString * _Nonnull)file function:(NSString * _Nonnull)function line:(int)line format:(NSString * _Nonnull)format, ... __attribute__((format(NSString, 5, 6)));
	// TODO: [Internal]
	[Abstract]
	[Export("tip_logWithLevel:file:function:line:format:", IsVariadic = true)]
	void Log(TIPLogLevel level, string file, string function, int line, string format, IntPtr varArgs);

	// @optional -(BOOL)tip_canLogWithLevel:(TIPLogLevel)level;
	[Export("tip_canLogWithLevel:")]
	bool CanLog(TIPLogLevel level);
}
```

```obj-c
@protocol TIPLogger <NSObject>

@required
- (void)tip_logWithLevel:(TIPLogLevel)level file:(nonnull NSString *)file function:(nonnull NSString *)function line:(int)line format:(nonnull NSString *)format, ... NS_FORMAT_FUNCTION(5,6);

@optional
- (BOOL)tip_canLogWithLevel:(TIPLogLevel)level;

@end
```

According to the docs: https://developer.xamarin.com/guides/cross-platform/macios/binding/objective-c-libraries/#Binding_Objective-C_Argument_Lists I can do stuff, but in this case it is a protocol that the USER must implement.

Any suggestions?

**Comment by migueldeicaza** [comment]
Your best in this scenario is to change the code in the Objective-C library to avoid calling the Log method with the variadic arguments and change the contract.   This effectively means we would need a fork of the library.

**Comment by mattleibow** [comment]
Thanks for confirming that this cannot be bound. I don't think we want to maintain a fork just yet, so we meet in the middle - I insert an extra file in the xcode project. And then we use that bindable type instead.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs [hunk 1]
```diff
@@ -0,0 +1,2147 @@
...
+		[Export("maxEstimatedTimeRemainingForDetachedHTTPDownloads")]
+		double MaxEstimatedTimeRemainingForDetachedHttpDownloads { get; set; }
+
+		// @property (copy, atomic) TIPEstimatedBitrateProviderBlock _Nullable estimatedBitrateProviderBlock;
+		[NullAllowed, Export("estimatedBitrateProviderBlock", ArgumentSemantic.Copy)]
+		TIPEstimatedBitrateProviderDelegate EstimatedBitrateProviderBlock { get; set; }
+
+		// @property (nonatomic) id<TIPImageFetchDownloadProvider> _Null_unspecified imageFetchDownloadProvider;
+		[Export("imageFetchDownloadProvider", ArgumentSemantic.Assign)]
+		TIPImageFetchDownloadProvider ImageFetchDownloadProvider { get; set; }
```
Comment (bholmes) [inline iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Not using the interface version of Protocols where possible.  See my diff for more instances.

https://gist.github.com/bholmes/12abcb7d4628f90e95dcc6b99beea834

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs [hunk 2]
```diff
@@ -0,0 +1,2147 @@
...
+
+	// @protocol TIPImageFetchDownloadClient <NSObject>
+	[Protocol, Model]
+	[BaseType(typeof(NSObject), Name = "TIPImageFetchDownloadClient")]
+	interface TIPImageFetchDownloadClient
+	{
+		// @required -(void)imageFetchDownloadDidStart:(id<TIPImageFetchDownload> _Nonnull)download;
+		[Abstract]
+		[Export("imageFetchDownloadDidStart:")]
+		void ImageFetchDownloadDidStart(ITIPImageFetchDownload download);
```
Comment (bholmes) [inline iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Consider dropping all of the ImageFetchDownload prefixes in the methods.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs [hunk 3]
```diff
@@ -0,0 +1,2147 @@
...
+	interface ITIPImageFetchDelegate { }
+
+	// @protocol TIPImageFetchDelegate <NSObject>
+	[Protocol, Model]
+	[BaseType(typeof(NSObject), Name = "TIPImageFetchDelegate")]
+	interface TIPImageFetchDelegate
+	{
+		// @optional -(void)tip_imageFetchOperationDidStart:(TIPImageFetchOperation * _Nonnull)op;
+		[Export("tip_imageFetchOperationDidStart:")]
+		void ImageFetchOperationDidStart(TIPImageFetchOperation op);
```
Comment (bholmes) [inline iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Consider dropping all of the ImageFetchOperation prefixes on all methods.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs [hunk 4]
```diff
@@ -0,0 +1,2147 @@
...
+		// @property (readonly, nonatomic) unsigned long long onDiskBytesUsed;
+		[Export("onDiskBytesUsed")]
+		ulong OnDiskBytesUsed { get; }
+	}
+
+	interface ITIPImagePipelineInspectionResultEntry { }
+
+	// @protocol TIPImagePipelineInspectionResultEntry <NSObject>
+	[Protocol, Model]
+	[BaseType(typeof(NSObject), Name = "TI
```
Comment (bholmes) [inline iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Consider removing all of the FetchHelper prefixes on all method names.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,2147 @@
...
+
+		// extern const NSTimeInterval TIPMaxEstimatedTimeRemainingForDetachedHTTPDownloadsDefault;
+		[Field("TIPMaxEstimatedTimeRemainingForDetachedHTTPDownloadsDefault", "__Internal")]
+		double MaxEstimatedTimeRemainingForDetachedHttpDownloads { get; }
+
+		// extern const NSUInteger TIPMaxRatioSizeOfCacheEntryDefault;
+		[Field("TIPMaxRatioSizeOfCacheEntryDefault", "__Internal")]
+		nuint MaxRatioSizeOfCacheEntry { get; }
+
+		// TODO: this is really a short
```
Comment (bholmes) [inline iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Did you want to change this?  Can it be a short?  There are 3 instances of this in this class.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs [hunk 6]
```diff
@@ -0,0 +1,2147 @@
...
+		string Identifier { get; }
+
+		// -(instancetype _Nullable)initWithIdentifier:(NSString * _Nonnull)identifier __attribute__((objc_designated_initializer));
+		[Export("initWithIdentifier:")]
+		[DesignatedInitializer]
+		IntPtr Constructor(string identifier);
+
+		// -(TIPImageFetchOperation * _Nonnull)operationWithRequest:(id<TIPImageFetchRequest> _Nonnull)request context:(id _Nullable)context delegate:(id<TIPImageFetchDelegate> _Nullable)delegate __attribute__((objc_requires_super));
+		[Export("operationWithRequest:context:delegate:")]
+		//[RequiresSuper] // TODO
```
Comment (bholmes) [inline iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: I am not 100% sure what to do here.  I am only getting one Google hit for this Attribute.  It looks like you should add an Advice attribute telling the user to call the base class?

https://bugzilla.xamarin.com/show_bug.cgi?id=58350

`[Advice ("You must call the base method when overriding.")]`

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/Additions.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+		public TIPImageFetchHydrationDelegate ImageRequestHydrationBlock { get; set; }
+
+		public NSUrl ImageUrl { get; set; }
+
+		public TIPImageFetchLoadingSources LoadingSources { get; set; }
+
+		public TIPImageFetchOptions Options { get; set; }
+
+		public NSDictionary<NSString, TIPImageFetchProgressiveLoadingPolicy> ProgressiveLoadingPolicies { get; set; }
```
Comment (bholmes) [inline iOS/TwitterImagePipeline/source/TwitterImagePipeline/Additions.cs]: Should be `ITIPImageFetchProgressiveLoadingPolicy `

File: manifest.yaml [hunk 8]
```diff
@@ -231,6 +231,10 @@
   BuildScript: ./iOS/LiquidFloatingActionButton/build.cake
   TriggerPaths: [ iOS/LiquidFloatingActionButton ]
   MacBuildTargets: [ samples, component ]
+- Name: TwitterImagePipeline
+  BuildScript: ./iOS/TwitterImagePipeline/build.cake
+  TriggerPaths: [ iOS/TwitterImagePipeline ]
+  MacBuildTargets: [ component, samples ]
```
Comment (bholmes) [inline manifest.yaml]: should be [ samples, component ]  

inconsistent order.

**Review by bholmes — CHANGES_REQUESTED** [review]
I have listed some comments for you to consider changing.  Also review all of your TODOs to ensure that you have not forgotten anything.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs (hunk 5 above)
Comment (mattleibow) [reply iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Currently, this is not yet supported by the binding engine: https://bugzilla.xamarin.com/show_bug.cgi?id=58845

But, I did it manually.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs (hunk 1 above)
Comment (mattleibow) [reply iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Updated my code.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs (hunk 6 above)
Comment (mattleibow) [reply iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: Done.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs (hunk 3 above)
Comment (mattleibow) [reply iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: done.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs (hunk 2 above)
Comment (mattleibow) [reply iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: done.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs (hunk 4 above)
Comment (mattleibow) [reply iOS/TwitterImagePipeline/source/TwitterImagePipeline/ApiDefinition.cs]: done.

File: iOS/TwitterImagePipeline/source/TwitterImagePipeline/Additions.cs (hunk 7 above)
Comment (mattleibow) [reply iOS/TwitterImagePipeline/source/TwitterImagePipeline/Additions.cs]: changed

File: manifest.yaml (hunk 8 above)
Comment (mattleibow) [reply manifest.yaml]: changed

**Review by bholmes — APPROVED** [review]

**Pull request merged**